automation/scripts/clean_pipelines/clean_pipelines.py

```diff
@@ -0,0 +1,392 @@
+            filter=query_filter if page_token == "" else "",
+        )
+        runs.extend([PipelineRun(run) for run in response.runs or []])
+        page_token = response.next_page_token
+        if counter % 50 == 0:
+            mlrun.utils.logger.info(
+                "Collecting pipelines runs:", runs_count=(10000 * counter)
+            )
+        counter += 1
+    return runs
+
+
+def query_and_filter_runs(
+    kfp_client: Client, project_name: str, query_filter: str
+) -> tuple[list[tuple[str, str]], set]:
+    """
+    Query the pipeline runs and filter them based on the project name.
+
+    :param kfp_client: KFP client for interacting with the pipeline API.
+    :param project_name: Name of the project for filtering the runs.
+    :param query_filter: Filter for querying the runs.
+    """
+    runs = list_pipelines_runs(kfp_client, query_filter)
+
+    # Filter out non project-related runs if project was provided
+    project_runs = (
+        filter_out_non_related_runs(project_name, runs) if project_name != "*" else runs
+    )
+    if project_name == "*":
+        project_names = [
+            mlrun_pipelines.mixins.PipelineProviderMixin().resolve_project_from_workflow_manifest(
+                run.workflow_manifest()
+            )
+            for run in project_runs
+        ]
+        project_names = set(project_names)
+    else:
+        project_names = {project_name}
+
+    mlrun.utils.logger.info(
+        f"Found {len(project_runs)} runs for projects", project_names=project_names
+    )
+    runs = [(run.id, run.name) for run in project_runs]
+
+    # Collect experiment IDs
+    experiment_ids = set(run.experiment_id for run in project_runs if run.experiment_id)
+
+    return runs, experiment_ids
+
+
+def find_empty_experiments(
+    kfp_client: Client, experiments_ids: set
+) -> list[tuple[str, str]]:
+    # Find empty experiments
+    empty_experiment_ids = []
+    for experiment_id in experiments_ids:
+        runs = kfp_client.list_runs(experiment_id=experiment_id)
+
+        if not runs.total_size:
+            experiment_name = get_experiment_name(kfp_client, experiment_id)
+            empty_experiment_ids.append((experiment_id, experiment_name))
+    return empty_experiment_ids
+
+
+def delete_runs(
+    context: mlrun.MLClientCtx,
+    kfp_client: Client,
+    runs: list[tuple[str, str]],
+    target_path: str,
+) -> None:
+    """
+    Delete pipeline runs based on the provided runs.
+
+    :param context: The context object to log results.
+    :param runs: List of tuples containing run IDs and names.
+    :param kfp_client: The KFP client used to interact with the pipeline API.
+    :param target_path: Path where details of successful and failed deletions will be logged as a dataset artifact
+    """
+    delete_items(
+        context,
+        runs,
+        lambda run_id: kfp_client._run_api.delete_run(run_id),
+        target_path=target_path,
+    )
+
+
+def delete_empty_experiments(
```

Comment:
In mlrun we don't care if the experiment is empty or not so why here?
This seems very heavy (doing a list for every experiment id) can you explain why we need it?
IIUC experiment and runs are 1 to 1 ? If I'm wrong then maybe kfp will block us from deleting a non-empty experiment and then we don't need to check?

Reply:
Actually, experiments can contain multiple runs, each time you execute the same pipeline under the same project, kfp creates a new run under the same experiment. This means an experiment can accumulate runs over time, and a project can have multiple experiments.

When we delete resources in MLRun for project deletion, we remove all assets associated with a project, including experiments created based on the project and workflow names. However, if we're deleting runs prior to a specific date, some runs within an experiment may remain. Deleting the entire experiment would remove runs that don't meet the deletion criteria. Therefore, we only need to delete an experiment if it becomes empty after all its associated runs have been deleted.